refactor!: custom flow into configurable oauth (#44)
famedly/synapse-token-authenticator: README.md

[RFC 7662 - OAuth 2.0 Token Introspection](https://datatracker.ietf.org/doc/html/rfc7662)
| Parameter          | Type                                                      |
|--------------------|-----------------------------------------------------------|
| `endpoint`         | String                                                    |
| `validator`        | [`Validator`](#Validator) (defaults to [`Exist`](#Exist)) |
| `auth`             | [`HttpAuth`](#HttpAuth) (optional)                        |
| `localpart_path`   | [`Path`](#Path) (optional)                                |
| `user_id_path`     | [`Path`](#Path) (optional)                                |
| `fq_uid_path`      | [`Path`](#Path) (optional)                                |
| `displayname_path` | [`Path`](#Path) (optional)                                |
| `required_scopes`  | Space separated string or a list of strings (optional)    |

Keep in mind, that default validator will always pass. According to the [spec](https://datatracker.ietf.org/doc/html/rfc7662), you probably want at least
```yaml
type: in
path: 'active'
validator:
  type: equal
  value: true
```
or
```yaml
['in', 'active', ['equal', true]]
```

> ### NotifyOnRegistration:
> | Parameter            | Type                               |
> |----------------------|------------------------------------|
> | `url`                | String                             |
> | `auth`               | [`HttpAuth`](#HttpAuth) (optional) |
> | `interrupt_on_error` | bool (defaults to `true`)          |
> 

Reviewer: ```suggestion
### NotifyOnRegistration:
| Parameter            | Type                               |
|----------------------|------------------------------------|
| `url`                | String                             |
| `auth`               | [`HttpAuth`](#HttpAuth) (optional) |
| `interrupt_on_error` | bool (defaults to `true`)          |

The notify on registration endpoint must return status code 200 with an empty response `{}`.
```

The notify on registration only works when the endpoint returns the status code 200 with an empty response `{}`. Therefore, I added this to the documentation.

Otherwise the STA fails, here are examples of other responses:

Status code 200 with no response:
```
synapse-1   | 2024-06-28 09:07:23,510 - synapse.http.client - 426 - INFO - POST-20 - Received response to POST https://apimocha.com/8b4537b2/register: 200
synapse-1   | 2024-06-28 09:07:23,511 - synapse.handlers.auth - 2179 - WARNING - POST-20 - Failed to run module API callback <bound method TokenAuthenticator.check_oauth of <synapse_token_authenticator.token_authenticator.TokenAuthenticator object at 0xffff96641130>>: Expecting value: line 1 column 1 (char 0)
```
Status code 201 example:
```
synapse-1   | 2024-06-28 08:51:13,902 - synapse.http.client - 426 - INFO - POST-4 - Received response to POST https://apimocha.com/8b4537b2/register: 201
synapse-1   | 2024-06-28 08:51:13,904 - synapse.handlers.auth - 2179 - WARNING - POST-4 - Failed to run module API callback <bound method TokenAuthenticator.check_oauth of <synapse_token_authenticator.token_authenticator.TokenAuthenticator object at 0xffffb1690350>>: Expecting value: line 1 column 1 (char 0)
```
Status code 204 example:
```
synapse-1   | 2024-06-28 09:02:41,162 - synapse.http.client - 426 - INFO - POST-4 - Received response to POST https://apimocha.com/8b4537b2/register: 204
synapse-1   | 2024-06-28 09:02:41,162 - synapse.handlers.auth - 2179 - WARNING - POST-4 - Failed to run module API callback <bound method TokenAuthenticator.check_oauth of <synapse_token_authenticator.token_authenticator.TokenAuthenticator object at 0xffff96641130>>: Expecting value: line 1 column 1 (char 0)
```

Author: Added an exception for value error: [synapse_token_authenticator/token_authenticator.py:481](https://github.com/famedly/synapse-token-authenticator/blob/92b1ff2ee44ef3bbada23b85a89d5a3f542dcba6/synapse_token_authenticator/token_authenticator.py#L481-L482). It should not fail on non-json responses now. I hope it works as expected now

Author: Also, changed the order for notification and registration: now it notifies first and registers after that. This way, with `interrupt_on_error` it won't register and authenticate if notify call failed

Reviewer: Sounds good! Thanks!

Authentication options, always optional
| Parameter | Type                    |
|-----------|-------------------------|
| `type`    | `'basic'` \| `'bearer'` |

Possible options: [`BasicAuth`](#BasicAuth), [`BearerAuth`](#BearerAuth),

### Validator
A validator is any of these types:
    [`Exist`](#Exist),
    [`Not`](#Not),
    [`Equal`](#Equal),
    [`MatchesRegex`](#MatchesRegex),
    [`AnyOf`](#AnyOf),
    [`AllOf`](#AllOf),
    [`In`](#In),
    [`ListAnyOf`](#ListAnyOf),
    [`ListAllOf`](#ListAllOf)

Each validator has `type` field

### Exist
Validator that always returns true

#### Examples
```yaml
{'type': 'exist'}
```
or
```yaml
['exist']
```

### Not
Validator that inverses the result of the inner validator

| Parameter   | Type                      |
|-------------|---------------------------|
| `validator` | [`Validator`](#Validator) |

#### Examples
```yaml
{'type': 'not', 'validator': 'exist'}
```
or
```yaml
['not', 'exist']
```

### Equal
Validator that checks for equality with a constant

| Parameter | Type  |
|-----------|-------|
| `value`   | `Any` |

#### Examples
```yaml
{'type': 'equal', 'value': 3}
```
or
```yaml
['equal', 3]
```

### MatchesRegex

> Validator that checks if a value is string and matches given regex

Reviewer 2: ```suggestion
Validator that checks if a value is string and matches the given regex
```

| Parameter                                  | Type   | Description                 |
|--------------------------------------------|--------|-----------------------------|
| `regex`                                    | `str`  | Python regex syntax         |
| `full_match` (optional, `true` by default) | `bool` | Full match or partial match |

#### Examples
```yaml
{'type': 'regex', 'regex': 'hello.'}
```
or
```yaml
['regex', 'hello.', false]
```

### AnyOf
Validator that checks if **any** of the inner validators passes

| Parameter    | Type                              |
|--------------|-----------------------------------|
| `validators` | List of [`Validator`](#Validator) |

#### Examples
```yaml
type: any_of
validators:
  - ['in', 'foo', ['equal', 3]]
  - ['in', 'bar' ['exist']]
```
or
```yaml
['any_of', [['in', 'bar' ['exist']], ['in', 'foo', ['equal', 3]]]]
```

### AllOf

> Validator that checks if **all** of the inner validators pass

Reviewer 2: No period at the end of the sentences here.

| Parameter    | Type                              |
|--------------|-----------------------------------|
| `validators` | List of [`Validator`](#Validator) |

#### Examples
```yaml
type: all_of
validators:
  - ['exist']
  - ['in', 'foo', ['equal', 3]]
```
or
```yaml
['all_of', [['exist'], ['in', 'foo', ['equal', 3]]]]
```

### In
Validator that modifies the context for the inner validator, *going inside* a dict key.

> If the validated object is not a dict, or doesn't have specivied `path`, validation failes. Validator

Reviewer 2: ```suggestion
If the validated object is not a dict, or doesn't have specified `path`, validation fails.
```

| Parameter   | Type                                                                |
|-------------|---------------------------------------------------------------------|
| `path`      | [`Path`](#Path)                                                     |
| `validator` | [`Validator`](#Validator) (optional, defaults to [`Exist`](#Exist)) |

#### Examples
```yaml
['in', ['foo', 'bar'], ['equal', 3]]
```

### ListAllOf

> Validator that checks if the value is list and **all** of its elements satisfy specified validator.

Reviewer 2: ```suggestion
Validator that checks if the value is list and **all** of its elements satisfy the specified validator.
```

| Parameter   | Type                      |
|-------------|---------------------------|
| `validator` | [`Validator`](#Validator) |

#### Examples
```yaml
type: list_all_of
validator:
  type: regex
  regex: 'ab..'
```
or
```yaml
['list_all_of', ['regex', 'ab..']]
```

### ListAnyOf

> Validator that checks if the value is list and if **any** of its elements satisfy specified validator.

Reviewer 2: ```suggestion
Validator that checks if the value is list and if **any** of its elements satisfy the specified validator.
```
